backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/TestingSession.java

```diff
@@ -0,0 +1,245 @@
+            this.behaviour = behaviour;
+        }
+
+        ResultSetFuture execute(Session session, Statement statement) {
+            return behaviour.apply(session, statement);
+        }
+    }
+
+    @FunctionalInterface
+    interface StatementPredicate extends Predicate<Statement> {
+
+    }
+
+    static class BoundStatementStartingWith implements StatementPredicate {
+        private final String queryStringPrefix;
+
+        BoundStatementStartingWith(String queryStringPrefix) {
+            this.queryStringPrefix = queryStringPrefix;
+        }
+
+        @Override
+        public boolean test(Statement statement) {
+            if (statement instanceof BoundStatement) {
+                BoundStatement boundStatement = (BoundStatement) statement;
+                return boundStatement.preparedStatement()
+                    .getQueryString()
+                    .startsWith(queryStringPrefix);
+            }
+            return false;
+        }
+    }
+
+    @FunctionalInterface
+    public interface RequiresCondition {
+        RequiresApplyCount condition(StatementPredicate statementPredicate);
+
+        default RequiresApplyCount always() {
+            return condition(ALL_STATEMENTS);
+        }
+
+        default RequiresApplyCount whenBoundStatementStartsWith(String queryStringPrefix) {
+            return condition(new BoundStatementStartingWith(queryStringPrefix));
+        }
+    }
+
+    @FunctionalInterface
+    public interface RequiresApplyCount {
+        FinalStage times(int applyCount);
+    }
+
+    @FunctionalInterface
+    public interface FinalStage {
+        void setExecutionHook();
+    }
+
+    private static class ExecutionHook {
+        final StatementPredicate statementPredicate;
+        final Behavior behavior;
+        final AtomicInteger remaining;
+
+        private ExecutionHook(StatementPredicate statementPredicate, Behavior behavior, int applyCount) {
+            this.statementPredicate = statementPredicate;
+            this.behavior = behavior;
+            this.remaining = new AtomicInteger(applyCount);
+        }
+
+        ResultSetFuture execute(Session session, Statement statement) {
+            if (statementPredicate.test(statement)) {
+                int hookPosition = remaining.getAndDecrement();
+                if (hookPosition > 0) {
+                    return behavior.execute(session, statement);
+                }
+            }
+            return Behavior.EXECUTE_NORMALLY.execute(session, statement);
+        }
+    }
+
+    private static StatementPredicate ALL_STATEMENTS = statement -> true;
+    private static ExecutionHook NO_EXECUTION_HOOK = new ExecutionHook(ALL_STATEMENTS, Behavior.EXECUTE_NORMALLY, 0);
+
+    private final Session delegate;
+    private volatile ExecutionHook executionHook;
+
+    TestingSession(Session delegate) {
+        this.delegate = delegate;
+        this.executionHook = NO_EXECUTION_HOOK;
+    }
+
+    public RequiresCondition fail() {
```

Comment:
we don't check that we are not overwriting an existing executionHook, I think that's a problem